functional/connectivity-loss_test.go

```diff
@@ -0,0 +1,229 @@
+	}
+	for name, command := range createUnits {
+		stdout, stderr, err := cluster.Fleetctl(m0, command...)
+		if err != nil {
+			t.Fatalf("Failed creating unit %s: %v\nstdout: %s\nstderr:%s", name, err, stdout, stderr)
+		}
+	}
+
+	checkExpectedStates := func() (isExpected bool, expected, actual map[string]string) {
+		// First check unit files.
+		// These shouldn't change at all after intital submit -- but better safe than sorry...
+		stdout, _, err := cluster.Fleetctl(m0, "list-unit-files", "--no-legend", "--full", "--fields", "unit,dstate")
+		stdout = strings.TrimSpace(stdout)
+		if err != nil {
+			t.Errorf("Failed listing unit files: %v", err)
+		}
+
+		lines := strings.Split(stdout, "\n")
+		actualUnitFiles := map[string]string{}
+		if stdout != "" {
+			for _, line := range lines {
+				cols := strings.Split(line, "\t")
+				actualUnitFiles[cols[0]] = cols[1]
+			}
+		}
+
+		if !reflect.DeepEqual(actualUnitFiles, expectedUnitFiles) {
+			return false, expectedUnitFiles, actualUnitFiles
+		}
+
+		// Now check the actual unit states.
+		stdout, _, err = cluster.Fleetctl(m0, "list-units", "--no-legend", "--full", "--fields", "unit,active")
+		stdout = strings.TrimSpace(stdout)
+		if err != nil {
+			t.Errorf("Failed listing units: %v", err)
+		}
+
+		lines = strings.Split(stdout, "\n")
+		actualUnitStates := map[string]string{}
+		if stdout != "" {
+			for _, line := range lines {
+				cols := strings.Split(line, "\t")
+				actualUnitStates[cols[0]] = cols[1]
+			}
+		}
+
+		return reflect.DeepEqual(actualUnitStates, expectedUnitStates), expectedUnitStates, actualUnitStates
+	}
+
+	// Wait for initial state being reached.
+	timeout := 15 * time.Second
+	alarm := time.After(timeout)
+	ticker := time.Tick(250 * time.Millisecond)
+loop:
+	for {
+		select {
+		case <-alarm:
+			t.Fatalf("Failed to reach expected initial state within %v.", timeout)
+		case <-ticker:
+			if isExpected, _, _ := checkExpectedStates(); isExpected {
+				break loop
+			}
+		}
+	}
+
+	// Cut connection to etcd.
+	//
+	// We use REJECT here, so fleet knows immediately that it's disconnected, rather than waiting for a timeout.
+	if _, err = cluster.MemberCommand(m0, "sudo", "iptables", "-I", "OUTPUT", "-p", "tcp", "-m", "multiport", "--dports=2379,4001", "-j", "REJECT"); err != nil {
+		t.Fatal(err)
+	}
+
+	// Wait long enough to be reasonably confident that no more state changes will happen.
+	ttl, _ := time.ParseDuration(util.FleetTTL)
+	agentReconcileInterval := 5 * time.Second
+	slack := 2 * time.Second
+
+	time.Sleep(ttl + agentReconcileInterval + slack)
+
+	// Check unit state after connection loss.
+	//
+	// Note: we cannot use fleetctl to check the state here,
+	// as fleet is not available to give us this information...
+	// We have to go deeper, and try to obtain the information from systemd directly.
+	stdout, err := cluster.MemberCommand(m0, "sudo", "sh", "-c", `'systemctl list-unit-files -t service --no-legend | ( grep "\(single\|global\)@" || true )'`)
```

Comment:
It is not necessary to use sudo to list unit files. Instead of grep you can use:

```sh
systemctl list-unit-files -t service --no-legend --no-pager single@.service global@.service
```